File: current_session/394.py
"""
394. Decode String

Given an encoded string, return it's decoded string.

The encoding rule is: k[encoded_string],
where the encoded_string inside the square brackets is being repeated exactly k times.
Note that k is guaranteed to be a positive integer.

You may assume that the input string is always valid; No extra white spaces,
square brackets are well-formed, etc.

Furthermore, you may assume that:
the original data does not contain any digits and that digits are only for those repeat numbers, k.
For example, there won't be input like 3a or 2[4].

Examples:
s = "3[a]2[bc]", return "aaabcbc".
s = "3[a2[c]]", return "accaccacc".
s = "2[abc]3[cd]ef", return "abcabccdcdcdef".
"""
class Solution(object):
    def decodeString(self, s):
        """
        :type s: str
        :rtype: str
        """
        # naive approach: go through from left to right
        # handle cases:
        # 1. non-digits
        # 2. digits
        # 3. brackets

        res = ""
        k = 0
        i = 0

        while i < len(s):
            if s[i] in "0123456789":
                k = 10*k + int(s[i])
                i += 1
            elif s[i] not in "[]":
                res += s[i]
                i += 1
            elif s[i] == "[":
                # find the correct right bracket
                j = self.finder(i, s)
                res += k*self.decodeString(s[i+1:j])
                i = j+1
                k = 0
            else:
                print "?", i, s[i]      # should not have this line printed though
                i += 1

        return res

    def finder(self, i, s):
        j = i+1
        need = 1
        while need > 0 and j < len(s):
            if s[j] == "[":
                need += 1
            elif s[j] == "]":
                need -= 1
            j += 1
        return j-1

print Solution().decodeString("3[a]2[bc]"), Solution().decodeString("3[a]2[bc]") == "aaabcbc"
print Solution().decodeString("3[a2[c]]"), Solution().decodeString("3[a2[c]]") == "accaccacc"
print Solution().decodeString("2[abc]3[cd]ef"), Solution().decodeString("2[abc]3[cd]ef") == "abcabccdcdcdef"

File: current_session/c++/394.cpp
#include <iostream>
#include <string>

using namespace std;

// clang++ -std=c++17 c++/394.cpp -o 394 && ./394 -v
class Solution {
public:
  string decodeString(string s) {
    string res;
    int k = 0;
    int i = 0;
    while (i < s.size()) {
      if (isdigit(s[i])) {
        k = 10*k + stoi(s.substr(i, 1));
        i++;
      } else if (s[i] == '[') {
        int j = finder(i, s);   // position of closing bracket
        // e.g. 'aa3[bcd]aa
        // .     01234567     -> j=7, i=3, l=j-i-1
        int l = j-i-1;
        string substr = decodeString(s.substr(i+1, l));
        for (int add = 0; add < k; add++) {
          res += substr;
        }
        i = j+1;
        k = 0;
      } else {
        res = res + s[i];
        i++;
      }
      // cout << "res now:  " << res << endl;
    }
    return res;
  }

  int finder(int start, string s) {
    int j = start+1;
    int need = 1;
    while (need > 0 & j < s.size()) {
      if (s[j] == '[') {
        need++;
      } else if (s[j] == ']') {
        need--;
      }
      j++;
    }
    // cout << "start" << start << " end" << j-1 << endl;
    return j-1;
  }
};

int main() {
  string s1 = "3[a]2[bc]";
  string s2 = "3[a2[c]]";
  string s3 = "2[abc]3[cd]ef";
  string s4 = "abc3[cd]xyz";
  cout << Solution().decodeString(s1) << endl;
  cout << Solution().decodeString(s2) << endl;
  cout << Solution().decodeString(s3) << endl;
  cout << Solution().decodeString(s4) << endl;
}